Add 0.4.19 release notes (#6376)
Napari/napari: docs/release/release_0_4_19.md

# napari 0.4.19

We're happy to announce the release of napari 0.4.19!
napari is a fast, interactive, multi-dimensional image viewer for Python.
It's designed for browsing, annotating, and analyzing large multi-dimensional
images. It's built on top of Qt (for the GUI), vispy (for performant GPU-based
rendering), and the scientific Python stack (numpy, scipy).

For more information, examples, and documentation, please visit our website:
https://github.com/napari/napari

## Highlights

> This release is mostly a bug fix release and hopefully the latest before all the excited features that are coming with version 0.5.0.

Reviewer: Small grammar catch.

```suggestion
This release is mostly a bug fix release and hopefully the latest before all the exciting features that are coming with version 0.5.0.
```

Author: please put suggestion here https://github.com/napari/napari-release-tools/pull/3

Reviewer: It's already covered by @brisvag's [suggestion](https://github.com/napari/napari-release-tools/pull/3/files#r1371474388)

# napari 0.4.19

We're happy to announce the release of napari 0.4.19!
napari is a fast, interactive, multi-dimensional image viewer for Python.
It's designed for browsing, annotating, and analyzing large multi-dimensional
images. It's built on top of Qt (for the GUI), vispy (for performant GPU-based
rendering), and the scientific Python stack (numpy, scipy).

For more information, examples, and documentation, please visit our website:

> https://github.com/napari/napari

Reviewer 2: ```suggestion
https://napari.org/stable/
```
?

Author: Thanks, I have update script for generation release notes.

# napari 0.4.19

We're happy to announce the release of napari 0.4.19!
napari is a fast, interactive, multi-dimensional image viewer for Python.
It's designed for browsing, annotating, and analyzing large multi-dimensional
images. It's built on top of Qt (for the GUI), vispy (for performant GPU-based
rendering), and the scientific Python stack (numpy, scipy).

For more information, examples, and documentation, please visit our website:
https://napari.org/stable/

## Highlights

> This release is mostly a bug fix release. We aim for the next release to be 0.5.0, with many new exciting features!

Reviewer 3: Just an idea to rephrase this to make it clear we're already stocked up on features for 0.5.0 :)

```suggestion
This release contains mostly a lot of bug fixes and improvements, as we prepare for the next, feature-rich release 0.5.0!
```

Highlights in `v0.4.19`:

A new feature: [napari/napari#6102](https://github.com/napari/napari/pull/6102) added convenient parsing of color names and hex color strings in layer data
and generating black-to-color colormaps when adding layers to the viewer. 
For example, one can now use: `viewer.add_image(data, colormap="crimson")` or `viewer.add_image(data, colormap="#FAFA00").
To see what color names can be handled, please refer to the 
[vispy` dictionary of colors](https://github.com/vispy/vispy/blob/269ed1ac4d8126421fd5a7eb06a2996d63f46b17/vispy/color/_color_dict.py#L181).
Note: there is a namespace collision for the case of `orange` which is both a proper `vispy` colormap name and a color name. 
Previously, `orange` was handled as the `vispy` colormap, which is from white to orange. 

> However, with 0.4.19, `orange` will be handled as a color and the generated colormap will be from black to orange. 

Reviewer 3: ```suggestion
**Colormaps**: [napari/napari#6102](https://github.com/napari/napari/pull/6102) added convenient parsing of color names and hex color strings in layer data
and generating black-to-color colormaps when adding layers to the viewer. 
For example, one can now use: `viewer.add_image(data, colormap="crimson")` or `viewer.add_image(data, colormap="#FAFA00").
To see what color names can be handled, please refer to the 
[vispy` dictionary of colors](https://github.com/vispy/vispy/blob/269ed1ac4d8126421fd5a7eb06a2996d63f46b17/vispy/color/_color_dict.py#L181).
Note: there is a namespace collision for the case of `orange` which is both a proper `vispy` colormap name and a color name. 
Previously, `orange` was handled as the `vispy` colormap, which is from white to orange. 
However, with 0.4.19, `orange` will be handled as a color and the generated colormap will be from black to orange. 
```

If you're interested in some feature that's currently behind private or deprecated API,
please make an issue on the napari repo requesting the feature to be exposed!

napari is now pydantic 2 compatible. At this moment we are using `pydantic.v1` to achieve this. 
Additionally, for this release, the napari conda bundle will be shipped with pydantic v1, but
we expect that in the next release we will ship the bundle with pydantic v2.
Please, if you are a plugin developer and your plugin uses pydantic, ensure
that it is compatible with pydantic v2 ([napari/napari/#6358](https://github.com/napari/napari/pull/6358)).

We have fixed problems with rendering some labels in incorrect color by sequence of Pull Requests:
[napari/napari/#3308](https://github.com/napari/napari/pull/3308),
[napari/napari/#6411](https://github.com/napari/napari/pull/6411),
[napari/napari/#6467](https://github.com/napari/napari/pull/6467),
[napari/napari/#6439](https://github.com/napari/napari/pull/6439).
[napari/napari/#6607](https://github.com/napari/napari/pull/6607).
This also improves performance of rendering labels.
However, there are still some limitations.
Best performance is for labels encoded as (u)int8.

> As GPU API accept 8 and 16 bits int textures and float textures,
> for labels stored as (u)int8 and (u)int16 we could transfer data to GPU without preprocessing.

Reviewer 4: ```suggestion
As the GPU API accepts 8 and 16 bits int textures and float textures,
for labels stored as (u)int8 and (u)int16 we could transfer data to GPU without preprocessing.
```

Author: could yo make remarks here https://github.com/napari/napari-release-tools/pull/3/files#diff-f175a32e4fcbdd7db7a328d3c6d0e2a6092f89a3480917f8e1d0fe50122ff026

Reviewer 4: Yes I will after meeting

Reviewer 3: Sorry, missed this. Will also move mine there.

> for labels stored as (u)int8 and (u)int16 we could transfer data to GPU without preprocessing.

Reviewer 3: ```suggestion
We have fixed several issues causing labels to be rendered with incorrect colors. It took several pull requests, and you can follow the process here:
[napari/napari/#3308](https://github.com/napari/napari/pull/3308),
[napari/napari/#6411](https://github.com/napari/napari/pull/6411),
[napari/napari/#6467](https://github.com/napari/napari/pull/6467),
[napari/napari/#6439](https://github.com/napari/napari/pull/6439).
[napari/napari/#6607](https://github.com/napari/napari/pull/6607).
As an added benefit, this also improved the rendering performance of labels.

To maximize performance, make sure to use the smallest possible data type that can fit your labels! (e.g: (u)int8, (u)int16). Otherwise, if you need bigger data types, installing the optional dependency `numba` may improve your performance.
```

> For bigger data types we need to convert data, so it leads to bigger memory usage.
> Also best performance for bigger data requires to install numba library.

Reviewer 4: ```suggestion
For bigger data types we need to convert data, which leads to bigger memory usage.
For best performance when working with bigger data, it is required to install the numba library.
```

> So if you wold like for best user performance, please try to use the minimal possible data type.
> Even if, in the future, we'll use more advanced techniques for rendering labels on GPU, that will allow
> transferring any data without preprocessing, the GPU often offers less VRAM that computer has RAM.
> Also we experiment a little with current options and increase the size of
> send data (uint8, uint16 and float32) decrease performance of rendering on GPU.
> The test was performed on NVIDIA GeForce RTX 3060 with 12GB VRAM.

Reviewer 4: ```suggestion
To achieve optimal user performance, it is advised to utilize the most compact data type possible.
Although future rendering techniques on the GPU may enable the direct transfer of data without pre-processing, it's important to consider that GPUs often have less VRAM compared to the system's RAM. 
Our experiments indicate that increasing the size of the sent data (such as using uint8, uint16, and float32) can negatively impact rendering performance on GPUs. These tests were performed on an NVIDIA GeForce RTX 3060 with 12GB VRAM.
```

> The test was performed on NVIDIA GeForce RTX 3060 with 12GB VRAM.

Reviewer 3: There's some info here that I don't think should be in the highlights. I would honestly just remove this whole chunk...

If you have any questions or suggestions regarding napari core, for example, 
on how to adapt your plugin to be compatible with future napari versions, please get in touch!
We have weekly community meetings, for which you can find the schedule [here](https://napari.org/stable/community/meeting_schedule.html).
We would like to hear from you.

- Use a shader for low discrepancy label conversion ([napari/napari/#3308](https://github.com/napari/napari/pull/3308))
- Automatic recognition of hex colour strings in layer data ([napari/napari/#6102](https://github.com/napari/napari/pull/6102))
- Postpone qt_viewer deprecation to 0.6.0 ([napari/napari/#6283](https://github.com/napari/napari/pull/6283))
- Pydantic 2 compatibility using `pydantic.v1`  ([napari/napari/#6358](https://github.com/napari/napari/pull/6358))

> - Initial deprecations for Labels API ([napari/napari/#6542](https://github.com/napari/napari/pull/6542))

Reviewer 3: Are these meant to be removed and/or moved to their corresponding paragraphs?

Author: Currently, they are automatically added by script. I'm not sure how to handle them properly. Maybe needs to be removed before merging.

Reviewer 3: Ah I see... No strong feelings for me, we can leave them in for simplicity.
